NEW shipment kits with dispatcher v2 (SQL)

lvessiller-opendsi opened the pull request:
NEW shipment kits with dispatcher v2 (SQL)
- repost #21034 for tables structure

- add "fk_product" in "expeditiondet" table (it exists since v20)

-  add "fk_parent" in "expeditiondet" table (it's for parent link of one product line to determine virtual product with children lines)

lvessiller-opendsi committed:
NEW shipment kits with dispatcher v2 (SQL)

lvessiller-opendsi committed:
Format sql comment

lvessiller-opendsi commented:
@eldy 
Do you think database structure is correct ?

lvessiller-opendsi committed:
NEW structure database for virtual products

lvessiller-opendsi committed:
Remove expedition det dispatcher tables

lvessiller-opendsi committed:
Update from develop

lvessiller-opendsi committed:
Update from develop

lvessiller-opendsi committed:
Update fk_product on expeditiondet table

lvessiller-opendsi committed:
Update from develop

lvessiller-opendsi committed:
Update from develop

lvessiller-opendsi committed:
Update from develop

lvessiller-opendsi committed:
Merge develop

lvessiller-opendsi committed:
Update from develop

lvessiller-opendsi committed:
Remove break line on sql migration file

lvessiller-opendsi committed:
Update left join for compatibility with Postgresql

lvessiller-opendsi committed:
Fix SQL join first table

lvessiller-opendsi committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into new-expedition-kits-with-dispatcher-sql2

lvessiller-opendsi committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into new-expedition-kits-with-dispatcher-sql2

lvessiller-opendsi committed:
Merge from develop

lvessiller-opendsi committed:
Merge on develop

eldy commented:
What is the goal of fk_parent for shipment ?
To manage the same feature that for order, proposal and invoice: grouping lines for subtotal ?
Or doing something else...

lvessiller-opendsi commented:
> What is the goal of fk_parent for shipment ? To manage the same feature that for order, proposal and invoice: grouping lines for subtotal ? Or doing something else...

The "fk_parent" is used for virtual product in a shipment line.
It's the id of parent line.
It's useful when you have virtual products and components in virtual product.
Such as : 
![image](https://github.com/user-attachments/assets/b70f59e6-e225-4013-8609-3d23bec168b9)

lvessiller-opendsi committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into new-expedition-kits-with-dispatcher-sql2

lvessiller-opendsi committed:
Resolve conflicts

lvessiller-opendsi committed:
Migration SQL for kits in v22

lvessiller-opendsi committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into new-expedition-kits-with-dispatcher-sql2

lvessiller-opendsi committed:
Merge from develop

lvessiller-opendsi committed:
Merge from develop

lvessiller-opendsi committed:
Merge from develop

lvessiller-opendsi commented:
@eldy 
It's ready to merge this SQL feature

lvessiller-opendsi committed:
Merge from develop

eldy commented:
I removed the migration line UPDATE llx_expeditiondet ...
Because when an expedition line is linked to a parent line (order line for example), the value of product to use must be the one of the parent. The fk_product can be used only if the line is linked to nothing.

eldy commented:
I replaced the update instruction in migration because update left join is not valid SQL. It may fails on some database or options.
I replaced with 
UPDATE llx_expeditiondet as ed SET ed.fk_product = (SELECT cd.fk_product FROM llx_commandedet as cd WHERE cd.rowid = ed.fk_elementdet AND ed.element_type = 'commande') AND ed.fk_product IS NULL;

Thanks to check it works as expected.

lvessiller-opendsi commented:
> I replaced the update instruction in migration because update left join is not valid SQL. It may fails on some database or options.
> 
> @lvessiller-opendsi Thanks to check it works as expected. I replaced with UPDATE llx_expeditiondet as ed SET ed.fk_product = (SELECT cd.fk_product FROM llx_commandedet as cd WHERE cd.rowid = ed.fk_elementdet AND ed.element_type = 'commande') WHERE ed.fk_product IS NULL;

For me this update is right.
